hack/lib/start.sh

```diff
@@ -0,0 +1,485 @@
+	wait_for_url "${API_SCHEME}://${API_HOST}:${API_PORT}/healthz/ready" "apiserver(ready): " 0.25 80
+	wait_for_url "${API_SCHEME}://${API_HOST}:${API_PORT}/api/v1/nodes/${KUBELET_HOST}" "apiserver(nodes): " 0.25 80
+
+	echo "[INFO] OpenShift server health checks done at: "
+	date
+}
+readonly -f os::start::internal::start_all_in_one
+
+# os::start::internal::start_etcd starts etcd for OpenShift
+# Globals:
+#  - USE_SUDO
+#  - LOG_DIR
+#  - MASTER_CONFIG_DIR
+#  - API_SCHEME
+#  - API_HOST
+#  - ETCD_PORT
+# Arguments:
+#  None
+# Returns:
+#  - export ETCD_PID
+function os::start::internal::start_etcd() {
+	local sudo="${USE_SUDO:+sudo}"
+
+	echo "[INFO] Starting etcd"
+	${sudo} env "PATH=${PATH}" OPENSHIFT_ON_PANIC=crash openshift start etcd \
+		--config=${MASTER_CONFIG_DIR}/master-config.yaml &>"${LOG_DIR}/etcd.log" &
+	export ETCD_PID=$!
+
+	echo "[INFO] etcd server start at: "
+	date 
+
+	wait_for_url "${API_SCHEME}://${API_HOST}:${ETCD_PORT}/version" "etcd: " 0.25 80
+
+	echo "[INFO] etcd server health checks done at: "
+	date
+}
+readonly -f os::start::internal::start_etcd
+
+# os::start::internal::start_api_server starts the OpenShift API server
+# Globals:
+#  - USE_SUDO
+#  - LOG_DIR
+#  - ARTIFACT_DIR
+#  - MASTER_CONFIG_DIR
+#  - API_SCHEME
+#  - API_HOST
+#  - API_PORT
+#  - KUBELET_SCHEME
+#  - KUBELET_HOST
+#  - KUBELET_PORT
+# Arguments:
+# 1 - api server version
+# Returns:
+#  - export OS_PID
+function os::start::internal::start_api_server() {
+	local sudo="${USE_SUDO:+sudo}"
+	local api_server_version=${1:-}
+
+	echo "[INFO] Starting OpenShift API server"
+	if [[ -n "${api_server_version}" ]]; then
+		${sudo} docker run -e OPENSHIFT_PROFILE=web -e OPENSHIFT_ON_PANIC=crash \
+			--privileged --net=host -v ${MASTER_CONFIG_DIR}:${MASTER_CONFIG_DIR} \
+			openshift/origin:${api_server_version} start master api \
+			--config=${MASTER_CONFIG_DIR}/master-config.yaml &>"${LOG_DIR}/apiserver.log" &
+	else
+		${sudo} env "PATH=${PATH}" OPENSHIFT_PROFILE=web OPENSHIFT_ON_PANIC=crash openshift start master api \
+			--config=${MASTER_CONFIG_DIR}/master-config.yaml \
+			&>"${LOG_DIR}/apiserver.log" &
+	fi
+	export API_SERVER_PID=$!
+
+	echo "[INFO] OpenShift API server start at: "
+	date
+
+	wait_for_url "${API_SCHEME}://${API_HOST}:${API_PORT}/healthz" "apiserver: " 0.25 80
+	wait_for_url "${API_SCHEME}://${API_HOST}:${API_PORT}/healthz/ready" "apiserver(ready): " 0.25 160
+
+	echo "[INFO] OpenShift API server health checks done at: "
+	date
+}
+readonly -f os::start::internal::start_api_server
+
+# os::start::internal::start_controllers starts the OpenShift controllers
+# Globals:
+#  - USE_SUDO
+#  - LOG_DIR
+#  - MASTER_CONFIG_DIR
+# Arguments:
+# 1 - controllers version
+# Returns:
+#  - export CONTROLLERS_PID
+function os::start::internal::start_controllers() {
+	local sudo="${USE_SUDO:+sudo}"
+	local controllers_version=${1:-}
+
+	echo "[INFO] Starting OpenShift controllers"
+	if [[ -n "${controllers_version}" ]]; then
```

Comment:
Two ways of doing this: one passes a "reference" to the array that the function can dereference from it's scope, the other just assigns values to special envars that the function knows to look for:
```sh
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail

function arrays_params() {
	local version="${1:-}"
	if [[ -z "${version}" ]]; then
		echo "openshift"
		return
	fi

	local volumes=
	local envars=
	if [[ "$#" -gt 1 ]]; then
		for volume in "${!2}"; do
			volumes+=" -v ${volume}:${volume}"
		done
	fi

	if [[ "$#" -gt 2 ]]; then
		for envar in "${!3}"; do
			envars+=" -e ${envar}"
		done
	fi

	echo "docker run ${volumes} ${envars} openshift/origin:${version}"
}

echo 'arrays as "parameters"'

echo no version
echo -n $'\t'
arrays_params

my_volumes=(
	'/var'
	'/proc'
)

my_envars=(
	'HOME=/home/myname'
	'GIT_EDITOR=vim'
)

echo no envars
echo -n $'\t'
arrays_params v1.2 my_volumes[@]

echo envars and vols
echo -n $'\t'
arrays_params v1.2 my_volumes[@] my_envars[@]


function arrays_envars() {
	local version="${1:-}"
	if [[ -z "${version}" ]]; then
		echo "openshift"
		return
	fi

	local volumes=
	local envars=
	if [[ -n "${volume_list:-}" ]]; then
		for volume in "${volume_list[@]}"; do
			volumes+=" -v ${volume}:${volume}"
		done
	fi

	if [[ -n "${envar_list:-}" ]]; then
		for envar in "${envar_list[@]}"; do
			envars+=" -e ${envar}"
		done
	fi

	echo "docker run ${volumes} ${envars} openshift/origin:${version}"
}

echo 'arrays as envars'

echo no version
echo -n $'\t'
arrays_envars

echo no envars
echo -n $'\t'
volume_list=( "${my_volumes[@]}" )
arrays_envars v1.2

echo envars and vols
echo -n $'\t'
envar_list=( "${my_envars[@]}" )
arrays_envars v1.2
```

I would choose the latter, the former is nice in that it's a little clearer that you're passing stuff through but it's a little magic for most people I'd bet.